keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,519 @@
+The administrative port serves the liveness probe and metrics.
+The liveness probe prevents a partially broken cluster
+where the KAS cannot connect to the cluster. This port also serves
+pprof debug commands and monitoring data for the proxy.
+
+### Direct Connection
+
+This connection type uses the default dialer.
+This allows use of the connectivity service without the connectivity proxy.
+This is a quick way to run the system in a “legacy” or fallback mode.
+Simple clusters (not needing network segregation) run this way to avoid the overhead
+(in latency or configuration) of the connectivity proxy.
+
+### Kubernetes API Server Outbound Requests
+
+The majority of the KAS communication originates from incoming requests.
+Here we cover the outgoing requests. This is our understanding of those requests
+and some details as to how they fit in this model. For the alpha release we
+support 'master', 'etcd' and 'cluster' connectivity service names.
+
+- **ETCD** It is possible to make etcd talk via the proxy.
+The etcd client takes a transport.
+(https://github.com/etcd-io/etcd/blob/master/client/client.go#L101)
+We will add configuration as to which proxy an etcd client should use.
+(https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go)
+This will add an extra process hop to our main scaling axis.
+We will scale test the impact and publish the results. As a precaution
+we will add an extra network configuration 'etcd' separate from ‘master’.
+Etcd requests can be configured separately from the rest of 'master'.
+- **Pods/Exec**, **Pods/Proxy**, **Pods/Portforward**, **Pods/Attach**, **Pods/Log**
+Pod requests (and pod sub-resource requests) are meant for the cluster
+and will be routed based on the ‘cluster’ NetworkContext.
+- **Nodes/Proxy**
+Node requests (and node sub-resource requests) are meant for the cluster
+and will be routed based on the ‘cluster’ NetworkContext.
+- **Services/Proxy**
+Service requests (and service sub-resource requests) are meant for the cluster
+and will be routed based on the ‘cluster’ NetworkContext.
+- **Admission Webhooks**
+Admission webhooks can be destined for a service or a URL.
+If destined for a service then the service rules apply (send to 'cluster').
+If destined for a URL then we will use the ‘master’ NetworkContext.
+- **Aggregated API Server (and OpenAPI requests for aggregated resources)**
+Aggregated API Servers can be destined for a service or a URL.
+If destined for a service then the service rules apply.
+If destined for a URL then we will use the ‘master’ NetworkContext.
+- **Authentication, Authorization and Audit Webhooks**
+These Webhooks use a kube config file to determine destination.
+Given that we use a ‘master’ NetworkContext.
+- **ImagePolicyWebhook**
+The image policy webhook uses a kube config file to determine destination.
+Given that we use a ‘master’ NetworkContext.
+- **KMS GRPC Service**
```

Comment:
:-) just a kudos on remembering this one, thanks!

Reply:
Thanks :smiley: